### PR TITLE
Adding keyword arguments to QubitDevice.execute and QubitDevice.apply

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
Currently the ``hash`` of a ``CircuitGraph`` is not available to devices.

**Description of the Change:**
Adds the opportunity to pass keyword argument to the ``QubitDevice.execute`` and ``QubitDevice.apply`` methods, such that the ``hash=circuit_graph.hash`` argument can be passed to these methods.

**Benefits:**
Devices will now be able to use these hashes for paramtric compilation capabilities.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: tests/test_qubit_device.py [hunk 1]
```diff
@@ -139,6 +155,29 @@ def test_op_queue_is_filled_during_execution(
         assert isinstance(call_history[2], qml.PauliZ)
         assert call_history[2].wires == [2]
 
+    def test_op_queue_is_filled_during_execution(
+        self, mock_qubit_device_with_paulis_and_methods, monkeypatch
+    ):
+        """Tests that the op_queue is correctly filled when pre_measure is called and that accessing
```
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_qubit_device.py]: Thanks, had a duplicated test case here!

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -175,6 +184,11 @@ def apply(self, operations, rotations=None, **kwargs):
             operations (list[~.Operation]): operations to apply to the device
             rotations (list[~.Operation]): operations that rotate the circuit
                 pre-measurement into the eigenbasis of the observables.
```
Comment (josh146) [inline pennylane/_qubit_device.py]: Need to remove rotations from here :)

File: pennylane/_qubit_device.py [hunk 3]
```diff
@@ -281,6 +295,14 @@ def states_to_binary(samples, num_wires):
         states_sampled_base_ten = samples[:, None] & powers_of_two
         return (states_sampled_base_ten > 0).astype(int)[:, ::-1]
 
+    @property
+    def circuit_hash(self):
+        """The hash of the circuit upon the last execution.
+
+        This can be used by devices in :meth:`apply` for parametric compilation.
+        """
+        return self._circuit_hash
```
Comment (josh146) [inline pennylane/_qubit_device.py]: Nice!

File: pennylane/_qubit_device.py [hunk 4]
```diff
@@ -281,6 +295,14 @@ def states_to_binary(samples, num_wires):
         states_sampled_base_ten = samples[:, None] & powers_of_two
         return (states_sampled_base_ten > 0).astype(int)[:, ::-1]
 
+    @property
+    def circuit_hash(self):
+        """The hash of the circuit upon the last execution.
+
+        This can be used by devices in :meth:`apply` for parametric compilation.
```
Comment (josh146) [inline pennylane/_qubit_device.py]:
```suggestion
        This can be used by devices in :meth:`~.apply` for parametric compilation.
```

**Pull request merged**